typed-racket-lib/typed-racket/private/type-contract.rkt

```diff
@@ -1435,116 +1440,103 @@
-        (if maybe-replacement+similar?
-          (let ()
-            (define replacement-sc (car maybe-replacement+similar?))
-            (define similar? (cdr maybe-replacement+similar?))
-            (define-values [similar-scs other-scs] (partition similar? rst))
-            (define new-fst (if (null? similar-scs) fst replacement-sc))
-            (cons new-fst (loop other-scs)))
-          (cons (car scs) (loop (cdr scs))))]))))
-
-(define merge-hash/sc*
-  (make-merge-overlapping-function
-    (λ (this-sc)
-      (match this-sc
-       [(or (hash/sc: this-K this-V)
-            (mutable-hash/sc: this-K this-V)
-            (immutable-hash/sc: this-K this-V)
-            (weak-hash/sc: this-K this-V))
-        (cons (hash/sc this-K this-V)
-              (λ (that-sc)
-                (match that-sc
-                 [(or (hash/sc: that-K that-V)
-                      (mutable-hash/sc: that-K that-V)
-                      (immutable-hash/sc: that-K that-V)
-                      (weak-hash/sc: that-K that-V))
-                  (and (equal? this-K that-K)
-                       (equal? this-V that-V))]
-                 [_ #false])))]
-       [_ #false]))))
-
-(define merge-vector/sc*
-  (make-merge-overlapping-function
-    (λ (this-sc)
-      (match this-sc
-       [(or (immutable-vector/sc: this-TS)
-            (mutable-vector/sc: this-TS)
-            (vector/sc: this-TS))
-        (cons (vector/sc this-TS)
-              (λ (that-sc)
-                (match that-sc
-                 [(or (immutable-vector/sc: that-TS)
-                      (mutable-vector/sc: that-TS)
-                      (vector/sc: that-TS))
-                  (equal? this-TS that-TS)]
-                 [_ #false])))]
-       [_ #false]))))
-
-(define merge-vectorof/sc*
-  (make-merge-overlapping-function
-    (λ (this-sc)
-      (match this-sc
-       [(or (immutable-vectorof/sc: this-T)
-            (mutable-vectorof/sc: this-T)
-            (vectorof/sc: this-T))
-        (cons (vectorof/sc this-T)
-              (λ (that-sc)
-                (match that-sc
-                 [(or (immutable-vectorof/sc: that-T)
-                      (mutable-vectorof/sc: that-T)
-                      (vectorof/sc: that-T))
-                  (equal? this-T that-T)]
-                 [_ #false])))]
-       [_ #false]))))
+(define (hash-type? ty)
+  (match ty
+   [(or (Mutable-HashTable: _ _)
+        (Mutable-HashTableTop:)
+        (Immutable-HashTable: _ _)
+        (Weak-HashTable: _ _)
+        (Weak-HashTableTop:))
+    #true]
+   [_ #false]))
+
+(define (vector-type? ty)
+  (match ty
+   [(or (Immutable-Vector: _)
+        (Mutable-Vector: _)
+        (Mutable-VectorTop:))
+    #true]
+   [_ #false]))
+
+(define (heterogeneous-vector-type? ty)
+  (match ty
+   [(or (Immutable-HeterogeneousVector: _)
+        (Mutable-HeterogeneousVector: _))
+    #true]
+   [_ #false]))
+
+(define (group-types-by-elems t->elems ts)
+  (group-by t->elems ts non-#f-and-equal?))
```

Comment:
Here & below, the code is more flexible that it needs to be. It lets `t->elems` return `#f` and calls the normal `(t->sc _)` if that happens.

If we raise an error instead of allowing `#f`, then we can delete this "group-types" function and one argument from `merge-types->scs`